File: 백준/Gold/9527. 1의 개수 세기/1의 개수 세기.py
from sys import stdin
def count_ones_in_binary_up_to(n):
    count = 0
    power_of_two = 1
    
    while power_of_two <= n:
        total_pairs = (n + 1) // (power_of_two * 2)
        remainder = (n + 1) % (power_of_two * 2)
        
        count += total_pairs * power_of_two
        count += max(0, remainder - power_of_two)
        
        power_of_two *= 2
        
    return count

def count_ones_between_A_and_B(A, B):
    return count_ones_in_binary_up_to(B) - count_ones_in_binary_up_to(A - 1)

# 예시 실행
A, B = map(int, stdin.readline().split())
result = count_ones_between_A_and_B(A, B)
print(result)

File: 백준/Gold/9527. 1의 개수 세기/1의 개수 세기.java
import java.util.*;
import java.io.*;

public class Main {

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        long A = sc.nextLong();
        long B = sc.nextLong();
        long result = countOnesBetweenAAndB(A, B);
        System.out.println(result);
    }

    public static long countOnesInBinaryUpTo(long n) {
        long count = 0;
        long powerOfTwo = 1;

        while (powerOfTwo <= n) {
            long totalPairs = (n + 1) / (powerOfTwo * 2);
            long remainder = (n + 1) % (powerOfTwo * 2);

            count += totalPairs * powerOfTwo;
            count += Math.max(0, remainder - powerOfTwo);

            powerOfTwo *= 2;
        }

        return count;
    }

    public static long countOnesBetweenAAndB(long A, long B) {
        return countOnesInBinaryUpTo(B) - countOnesInBinaryUpTo(A - 1);
    }
}